Jax Einsum Test fails due to `complex.constant`

It appears IREE does not know how to generate buffers from `complex.constant` values.  See attach IR snippet

```
#map = affine_map<(d0, d1, d2) -> (d2, d0, d1)>
#map1 = affine_map<(d0, d1, d2) -> (d0, d1)>
module @jit__einsum {
  func.func public @main(%arg0: tensor<2x2x3xcomplex<f32>> {mhlo.sharding = "{replicated}"}, %arg1: tensor<3x4xcomplex<f32>> {mhlo.sharding = "{replicated}"}) -> (tensor<2x3xcomplex<f32>> {jax.result_info = ""}) {
    %cst = complex.constant [0.000000e+00 : f32, 0.000000e+00 : f32] : complex<f32>
    %cst_0 = arith.constant 0.000000e+00 : f32
    %0 = tensor.empty() : tensor<2x3xcomplex<f32>>
    %1 = linalg.fill ins(%cst : complex<f32>) outs(%0 : tensor<2x3xcomplex<f32>>) -> tensor<2x3xcomplex<f32>>
    %2 = linalg.generic {indexing_maps = [#map, #map1], iterator_types = ["parallel", "parallel", "reduction"]} ins(%arg0 : tensor<2x2x3xcomplex<f32>>) outs(%1 : tensor<2x3xcomplex<f32>>) {
    ^bb0(%in: complex<f32>, %out: complex<f32>):
      %3 = linalg.index 0 : index
      %4 = linalg.index 2 : index
      %5 = arith.index_cast %3 : index to i32
      %6 = arith.index_cast %4 : index to i32
      %7 = arith.cmpi eq, %6, %5 : i32
      %8 = arith.uitofp %7 : i1 to f32
      %9 = complex.create %8, %cst_0 : complex<f32>
      %10 = complex.mul %in, %9 : complex<f32>
      %11 = complex.add %out, %10 : complex<f32>
      linalg.yield %11 : complex<f32>
    } -> tensor<2x3xcomplex<f32>>
    return %2 : tensor<2x3xcomplex<f32>>
  }
}
```

with error:

```
lax_numpy_einsum_test.py:37:0: error: 'stream.async.dispatch' op operand #3 must be resource or external resource or transient resource or variable resource or constant resource or index or integer or floating-point, but got 'complex<f32>'
lax_numpy_einsum_test.py:37:0: note: see current operation: %10 = "stream.async.dispatch"(%0, %2, %1, %4, %arg0, %7, %arg1, %6, %8, %8, %arg1, %6, %arg1, %6, %9) {entry_point = @_main_dispatch_1::@_main_dispatch_1_generic_2x3x2, operand_segment_sizes = array<i32: 3, 3, 2, 2, 2, 2, 1>, tied_operands = [-1 : index]} : (index, index, index, complex<f32>, !stream.resource<*>, !stream.resource<*>, index, index, index, index, index, index, index, index, index) -> !stream.resource<*>
```

--- a/compiler/src/iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp
+++ b/compiler/src/iree/compiler/Dialect/Flow/Transforms/OutlineDispatchRegions.cpp
@@ -11,6 +11,7 @@
 #include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
 #include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
 #include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "iree/compiler/Utils/StringUtils.h"
 #include "llvm/ADT/StringExtras.h"
 #include "llvm/Support/Debug.h"
 #include "mlir/Dialect/Func/IR/FuncOps.h"
@@ -256,6 +257,9 @@ static std::string summarizeDispatchWorkgroupsOp(
         // No summarization implemented, default to the op's name.
         bestSummary = op->getName().getStringRef().str();
       });
+
+  // Sanitize the string so that it contains only C literal-compatible chars.
+  bestSummary = sanitizeSymbolName(bestSummary);
 
   LLVM_DEBUG(llvm::dbgs() << "// best op summary: '" << bestSummary << "'\n");
   return bestSummary;

--- a/compiler/src/iree/compiler/Dialect/Stream/Transforms/PackDispatchOperands.cpp
+++ b/compiler/src/iree/compiler/Dialect/Stream/Transforms/PackDispatchOperands.cpp
@@ -11,6 +11,7 @@
 #include "iree/compiler/Dialect/Stream/Transforms/Passes.h"
 #include "llvm/Support/Debug.h"
 #include "mlir/Dialect/Arith/IR/Arith.h"
+#include "mlir/Dialect/Complex/IR/Complex.h"
 #include "mlir/Dialect/Func/IR/FuncOps.h"
 #include "mlir/IR/Attributes.h"
 #include "mlir/IR/Builders.h"
@@ -62,8 +63,78 @@ namespace {
 // any i64 (or equivalent) operand and splits it in to two i32s that get
 // stitched back together.
 
-// TODO(benvanik): handle index types here based on target devices. For now
-// we assume 32-bit device sizes and cast all index operands to i32.
+// Converts |operand| to a sequence of one or more i32 values.
+static void convertAndDecomposeToI32s(
+    Location loc, Value operand, SmallVector<Value> &newOperands,
+    IREE::Stream::ResourceConfigAttr resourceConfig, OpBuilder &builder) {
+  // NOTE: we do these in sequence so that we can reuse type expansion; we
+  // want f64 to become i64 so that i64 can become i32+i32 etc.
+
+  // Complex types are decomposed into real/imaginary components and processed
+  // as if they were independent.
+  if (auto complexType = dyn_cast<ComplexType>(operand.getType())) {
+    auto real = builder.create<complex::ReOp>(loc, operand);
+    auto imag = builder.create<complex::ImOp>(loc, operand);
+    convertAndDecomposeToI32s(loc, real, newOperands, resourceConfig, builder);
+    convertAndDecomposeToI32s(loc, imag, newOperands, resourceConfig, builder);
+    return;
+  }
+
+  // index -> i32 or i64
+  if (operand.getType().isIndex()) {
+    // Convert index types to a concrete bit width. 'index' can mean different
+    // things on the host and device as well as varying across devices.
+    // Today we just hardcode to i32 as we are working with smallish data and
+    // i64 uses 2x as much of our limited push constant space and is much
+    // slower to work with on mobile GPUs. In the future we will want to flag
+    // this as a global setting as well as have some heuristics for deriving
+    // from target devices.
+    operand = builder.createOrFold<arith::IndexCastUIOp>(
+        loc, builder.getIntegerType(resourceConfig.getIndexBits()), operand);
+  }
+
+  // bf16 -> i16, f32 -> i32, f64 -> i64 ...
+  if (auto floatType = dyn_cast<FloatType>(operand.getType())) {
+    // Floats need to be bitcast into opaque integers so that our handling
+    // below can deal with simple fixed width ints (bf16->i16, etc).
+    operand = builder.createOrFold<arith::BitcastOp>(
+        loc, builder.getIntegerType(floatType.getIntOrFloatBitWidth()),
+        operand);
+  }
+
+  // i1-i31 -> i32 and i33-i63 -> i64
+  // TODO(benvanik): don't extend here but instead pack as we can fit 4 i8's
+  // into a single i32 and save 4x our push constant capacity
+  unsigned bitWidth = IREE::Util::getTypeBitWidth(operand.getType());
+  if (bitWidth < 31) {
+    operand = builder.createOrFold<arith::ExtUIOp>(loc, builder.getI32Type(),
+                                                   operand);
+  } else if (bitWidth > 32 && bitWidth < 64) {
+    operand = builder.createOrFold<arith::ExtUIOp>(loc, builder.getI64Type(),
+                                                   operand);
+  }
+
+  // i64 -> i32 + i32
+  if (operand.getType().isInteger(64)) {
+    // TODO(benvanik): reorder operands to preserve natural alignment; right
+    // now this can split i64 across 4 byte boundaries which sucks.
+    // TODO(benvanik): use something like IndexSet for the shift amount to
+    // reduce the amount of IR we emit when passing multiple operands.
+    // lo = i32(operand)
+    // hi = i32(operand >> 32)
+    auto lo = builder.createOrFold<arith::TruncIOp>(loc, builder.getI32Type(),
+                                                    operand);
+    auto hi = builder.createOrFold<arith::TruncIOp>(
+        loc, builder.getI32Type(),
+        builder.createOrFold<arith::ShRUIOp>(
+            loc, builder.getI64Type(), operand,
+            builder.create<arith::ConstantIntOp>(loc, 32, 64)));
+    newOperands.push_back(lo);
+    newOperands.push_back(hi);
+  } else {
+    newOperands.push_back(operand);
+  }
+}
 
 // Converts stream.cmd.dispatch operands into their packed representation. This
 // will add/remove/reorder operands and must mirrored in a consistent manner
@@ -80,65 +151,97 @@ static void updateDispatchOp(IREE::Stream::CmdDispatchOp dispatchOp,
   auto loc = dispatchOp.getLoc();
   SmallVector<Value> newOperands;
   for (auto operand : dispatchOp.getUniformOperands()) {
-    // NOTE: we do these in sequence so that we can reuse type expansion; we
-    // want f64 to become i64 so that i64 can become i32+i32 etc.
-
-    // index -> i32 or i64
-    if (operand.getType().isIndex()) {
-      // Convert index types to a concrete bit width. 'index' can mean different
-      // things on the host and device as well as varying across devices.
-      // Today we just hardcode to i32 as we are working with smallish data and
-      // i64 uses 2x as much of our limited push constant space and is much
-      // slower to work with on mobile GPUs. In the future we will want to flag
-      // this as a global setting as well as have some heuristics for deriving
-      // from target devices.
-      operand = builder.createOrFold<arith::IndexCastUIOp>(
-          loc, builder.getIntegerType(resourceConfig.getIndexBits()), operand);
-    }
-
-    // bf16 -> i16, f32 -> i32, f64 -> i64 ...
-    if (auto floatType = llvm::dyn_cast<FloatType>(operand.getType())) {
-      // Floats need to be bitcast into opaque integers so that our handling
-      // below can deal with simple fixed width ints (bf16->i16, etc).
-      operand = builder.createOrFold<arith::BitcastOp>(
-          loc, builder.getIntegerType(floatType.getIntOrFloatBitWidth()),
-          operand);
-    }
-
-    // i1-i31 -> i32 and i33-i63 -> i64
-    // TODO(benvanik): don't extend here but instead pack as we can fit 4 i8's
-    // into a single i32 and save 4x our push constant capacity
-    unsigned bitWidth = IREE::Util::getTypeBitWidth(operand.getType());
-    if (bitWidth < 31) {
-      operand = builder.createOrFold<arith::ExtUIOp>(loc, builder.getI32Type(),
-                                                     operand);
-    } else if (bitWidth > 32 && bitWidth < 64) {
-      operand = builder.createOrFold<arith::ExtUIOp>(loc, builder.getI64Type(),
-                                                     operand);
-    }
-
-    // i64 -> i32 + i32
-    if (operand.getType().isInteger(64)) {
-      // TODO(benvanik): reorder operands to preserve natural alignment; right
-      // now this can split i64 across 4 byte boundaries which sucks.
-      // TODO(benvanik): use something like IndexSet for the shift amount to
-      // reduce the amount of IR we emit when passing multiple operands.
-      // lo = i32(operand)
-      // hi = i32(operand >> 32)
-      auto lo = builder.createOrFold<arith::TruncIOp>(loc, builder.getI32Type(),
-                                                      operand);
-      auto hi = builder.createOrFold<arith::TruncIOp>(
-          loc, builder.getI32Type(),
-          builder.createOrFold<arith::ShRUIOp>(
-              loc, builder.getI64Type(), operand,
-              builder.create<arith::ConstantIntOp>(loc, 32, 64)));
-      newOperands.push_back(lo);
-      newOperands.push_back(hi);
-    } else {
-      newOperands.push_back(operand);
-    }
+    // Decompose the operand to a sequence of i32 values.
+    // This must be consistent with recomposeFromI32s.
+    convertAndDecomposeToI32s(loc, operand, newOperands, resourceConfig,
+                              builder);
   }
   dispatchOp.getUniformOperandsMutable().assign(newOperands);
+}
+
+// Recompose integer values from multiple i64 values. The values remain in their
+// integer types and need to be reconverted.
+static Value recomposeFromI32sAndConvert(
+    Block *entryBlock, Location loc, Type oldArgType, DictionaryAttr oldArgAttr,
+    SmallVector<Type> &newArgTypes, SmallVector<DictionaryAttr> &newArgAttrs,
+    IREE::Stream::ResourceConfigAttr resourceConfig, OpBuilder &builder) {
+  // NOTE: we do the packing in a reverse sequence so that we can reuse type
+  // expansion; if we did f64->i64->i32+i32 in convertAndDecomposeToI32s we need
+  // to do i32+i32->i64->f64 here.
+
+  // Complex types need to be reconstructed from their elemental parts.
+  if (auto complexType = dyn_cast<ComplexType>(oldArgType)) {
+    Value real = recomposeFromI32sAndConvert(
+        entryBlock, loc, complexType.getElementType(), oldArgAttr, newArgTypes,
+        newArgAttrs, resourceConfig, builder);
+    Value imag = recomposeFromI32sAndConvert(
+        entryBlock, loc, complexType.getElementType(), oldArgAttr, newArgTypes,
+        newArgAttrs, resourceConfig, builder);
+    return builder.create<complex::CreateOp>(loc, oldArgType, real, imag);
+  }
+
+  // Pass through/no change other types (!stream.binding, probably).
+  if (!oldArgType.isIntOrIndexOrFloat()) {
+    newArgTypes.push_back(oldArgType);
+    newArgAttrs.push_back(oldArgAttr);
+    return entryBlock->addArgument(oldArgType, loc);
+  }
+
+  // If the arg was decomposed into i32s then first recompose it into an i64.
+  Value value;
+  bool wasDecomposed =
+      (oldArgType.isIndex() && resourceConfig.getIndexBits() > 32) ||
+      (oldArgType.isIntOrFloat() && oldArgType.getIntOrFloatBitWidth() > 32);
+  if (wasDecomposed) {
+    // The existing arg becomes the lo word and we need to insert the hi word.
+    auto loArg = entryBlock->addArgument(builder.getI32Type(), loc);
+    newArgTypes.push_back(loArg.getType());
+    auto hiArg = entryBlock->addArgument(builder.getI32Type(), loc);
+    newArgTypes.push_back(hiArg.getType());
+    // i64(lo) | (i64(hi) << 32)
+    value = builder.create<arith::OrIOp>(
+        loc, builder.create<arith::ExtUIOp>(loc, builder.getI64Type(), loArg),
+        builder.create<arith::ShLIOp>(
+            loc,
+            builder.create<arith::ExtUIOp>(loc, builder.getI64Type(), hiArg),
+            builder.create<arith::ConstantIntOp>(loc, 32, 64)));
+  } else {
+    // Forced bitcast.
+    value = entryBlock->addArgument(builder.getI32Type(), loc);
+    newArgTypes.push_back(value.getType());
+  }
+
+  // i32 or i64 -> index
+  if (oldArgType.isIndex()) {
+    value = builder.create<arith::IndexCastUIOp>(loc, builder.getIndexType(),
+                                                 value);
+  }
+
+  // Truncate back to original bit width.
+  // i32 -> i16, i64 -> i48, ...
+  if (oldArgType.isIntOrFloat() && oldArgType.getIntOrFloatBitWidth() < 32) {
+    value = builder.create<arith::TruncIOp>(
+        loc, builder.getIntegerType(oldArgType.getIntOrFloatBitWidth()), value);
+  }
+
+  // i16 -> bf16, i32 -> f32, i64 -> f64 ...
+  if (auto floatType = llvm::dyn_cast<FloatType>(oldArgType)) {
+    value = builder.create<arith::BitcastOp>(loc, oldArgType, value);
+  }
+
+  // Preserve the arg attrs on either the final op or the function argument
+  // if none was required.
+  if (auto definingOp = value.getDefiningOp()) {
+    if (oldArgAttr) definingOp->setAttrs(oldArgAttr);
+    newArgAttrs.push_back(nullptr);
+  } else {
+    newArgAttrs.push_back(oldArgAttr);
+  }
+  // Note that if we had decomposed the arg we'll expect that there are two attr
+  // dicts for the two new args.
+  if (wasDecomposed) newArgAttrs.push_back(nullptr);
+
+  return value;
 }
 
 // Updates an exported function in a stream.executable to match the packing
@@ -149,164 +252,29 @@ static void updateExportFuncOp(mlir::func::FuncOp funcOp) {
   assert(!funcOp.empty() && "can't have empty exported functions");
   auto &entryBlock = funcOp.getFunctionBody().front();
   auto builder = OpBuilder::atBlockBegin(&entryBlock);
-  auto streamAlignmentAttr = builder.getStringAttr("stream.alignment");
-  auto streamValuesAttr = builder.getStringAttr("stream.values");
 
   auto resourceConfig = IREE::Stream::ResourceConfigAttr::lookup(funcOp);
 
-  // NOTE: we have !stream.binding mixed in here; we only want to look at
-  // primitives.
-
-  // NOTE: we do the packing in a reverse sequence so that we can reuse type
-  // expansion; if we did f64->i64->i32+i32 above we need to do
-  // i32+i32->i64->f64 here.
-
-  // i32+i32->i64 is the only case that adds arguments today so to keep things
-  // simple we do that first. This ensures we have single SSA values for arg
-  // to operand mapping.
+  // Recompose i32 args into i64s and (if needed) convert them.
+  // This appends new arguments and then we clean up the old ones below.
   SmallVector<Type> newArgTypes;
   SmallVector<DictionaryAttr> newArgAttrs;
-  auto oldArgAttrs = funcOp.getAllArgAttrs();
-  for (auto it :
-       llvm::enumerate(llvm::to_vector<4>(funcOp.getArgumentTypes()))) {
-    auto targetType = it.value();
-    if (!targetType.isIntOrIndexOrFloat() ||
-        (targetType.isIndex() && resourceConfig.getIndexBits() <= 32) ||
-        (targetType.isIntOrFloat() &&
-         targetType.getIntOrFloatBitWidth() <= 32)) {
-      // Pass through/no change.
-      if (oldArgAttrs) {
-        newArgAttrs.push_back(
-            llvm::dyn_cast_if_present<DictionaryAttr>(oldArgAttrs[it.index()]));
-      } else {
-        newArgAttrs.push_back(nullptr);
-      }
-      newArgTypes.push_back(targetType);
-      continue;
-    }
-
-    // The existing arg becomes the lo word and we need to insert the hi word.
-    auto loArg = entryBlock.getArgument(newArgTypes.size());
-    newArgTypes.push_back(builder.getI32Type());
-    loArg.setType(builder.getI32Type());
-    auto hiArg = entryBlock.insertArgument(
-        newArgTypes.size(), builder.getI32Type(), funcOp.getLoc());
-    newArgTypes.push_back(builder.getI32Type());
-
-    // i64(lo) | (i64(hi) << 32)
-    auto loc = loArg.getLoc();
-    auto loArgI64 =
-        builder.create<arith::ExtUIOp>(loc, builder.getI64Type(), loArg);
-    Operation *argOp = builder.create<arith::OrIOp>(
-        loc, loArgI64,
-        builder.create<arith::ShLIOp>(
-            loc,
-            builder.create<arith::ExtUIOp>(loc, builder.getI64Type(), hiArg),
-            builder.create<arith::ConstantIntOp>(loc, 32, 64)));
-
-    // If going to index we need to insert a cast as the type changes.
-    if (targetType.isIndex()) {
-      argOp = builder.create<arith::IndexCastUIOp>(loc, targetType,
-                                                   argOp->getResult(0));
-    }
-
-    // Replace all subsequent uses with the new reconstituted value.
-    loArg.replaceAllUsesExcept(argOp->getResult(0), loArgI64);
-
-    // Take the existing potential values, if present, and fix them up.
-    // If we had a `stream.values = [0xAA...BB... : i64]` then we can
-    // put a `stream.values = [0xBB... : i32]` on the lo and `[0xAA... : i32]`
-    // on the hi.
-    //
-    // TODO(benvanik): find a nicer way/write some utilities - manipulating
-    // these attributes is really annoying.
-    if (oldArgAttrs) {
-      auto oldArgAttr =
-          llvm::dyn_cast_if_present<DictionaryAttr>(oldArgAttrs[it.index()]);
-      if (auto alignmentAttr =
-              oldArgAttr.getAs<IntegerAttr>(streamAlignmentAttr)) {
-        argOp->setAttr(streamAlignmentAttr, alignmentAttr);
-      }
-      if (auto valuesAttr = oldArgAttr.getAs<ArrayAttr>(streamValuesAttr)) {
-        // Attach the original potential value set to the new arg value.
-        // This allows any further analysis walking the SSA values to find the
-        // easily understandable i64 types instead of having to reconstruct it
-        // from the i32 values and the combining operations (which MLIR doesn't
-        // do today).
-        argOp->setAttr(streamValuesAttr, valuesAttr);
-      }
-    }
-    newArgAttrs.push_back(nullptr);
-    newArgAttrs.push_back(nullptr);
-  }
-  if (newArgTypes.size() != funcOp.getNumArguments()) {
-    // Changed argument count; update signature.
-    funcOp.setType(builder.getFunctionType(
-        newArgTypes, funcOp.getFunctionType().getResults()));
-    funcOp.setAllArgAttrs(newArgAttrs);
+  auto oldArgs = llvm::to_vector(entryBlock.getArguments());
+  for (auto oldArg : oldArgs) {
+    auto oldArgAttr = funcOp.getArgAttrDict(oldArg.getArgNumber());
+    auto newArg = recomposeFromI32sAndConvert(
+        &entryBlock, oldArg.getLoc(), oldArg.getType(), oldArgAttr, newArgTypes,
+        newArgAttrs, resourceConfig, builder);
+    oldArg.replaceAllUsesWith(newArg);
   }
 
-  // -- now the function has all <= 32-bit operands --
+  // Remove all the original arguments from the entry block.
+  entryBlock.eraseArguments(0, oldArgs.size());
 
-  newArgTypes.clear();
-  for (auto arg : entryBlock.getArguments()) {
-    auto alignmentAttr = funcOp.getArgAttrOfType<IntegerAttr>(
-        arg.getArgNumber(), streamAlignmentAttr);
-    auto oldValuesAttr = funcOp.getArgAttrOfType<ArrayAttr>(arg.getArgNumber(),
-                                                            streamValuesAttr);
-    auto targetType = arg.getType();
-    if (!targetType.isIntOrIndexOrFloat()) {
-      // Pass-through !stream.bindings.
-      newArgTypes.push_back(targetType);
-      continue;
-    }
-
-    auto loc = arg.getLoc();
-    Value value = arg;
-    arg.setType(builder.getI32Type());
-
-    // i32 or i64 -> index
-    if (targetType.isIndex()) {
-      auto castOp = builder.create<arith::IndexCastUIOp>(
-          loc, builder.getIndexType(), value);
-      value.replaceAllUsesExcept(castOp.getResult(), castOp);
-      value = castOp.getResult();
-    }
-
-    // Truncate back to original bit width.
-    // i32 -> i16, i64 -> i48, ...
-    if (targetType.isIntOrFloat() && targetType.getIntOrFloatBitWidth() < 32) {
-      auto truncOp = builder.create<arith::TruncIOp>(
-          loc, builder.getIntegerType(targetType.getIntOrFloatBitWidth()),
-          value);
-      value.replaceAllUsesExcept(truncOp.getResult(), truncOp);
-      value = truncOp.getResult();
-    }
-
-    // i16 -> bf16, i32 -> f32, i64 -> f64 ...
-    if (auto floatType = llvm::dyn_cast<FloatType>(targetType)) {
-      auto bitcastOp = builder.create<arith::BitcastOp>(loc, targetType, value);
-      value.replaceAllUsesExcept(bitcastOp.getResult(), bitcastOp);
-      value = bitcastOp.getResult();
-    }
-
-    // Set the original streams.values attribute on the op with the type in the
-    // original form. This allows subsequent analysis to easily find the value.
-    if (auto definingOp = value.getDefiningOp()) {
-      if (alignmentAttr) {
-        definingOp->setAttr(streamAlignmentAttr, alignmentAttr);
-        funcOp.removeArgAttr(arg.getArgNumber(), streamAlignmentAttr);
-      }
-      if (oldValuesAttr) {
-        definingOp->setAttr(streamValuesAttr, oldValuesAttr);
-        funcOp.removeArgAttr(arg.getArgNumber(), streamValuesAttr);
-      }
-    }
-
-    newArgTypes.push_back(builder.getI32Type());
-  }
+  // Update the function signature and arg attrs that may have changed.
   funcOp.setType(builder.getFunctionType(
       newArgTypes, funcOp.getFunctionType().getResults()));
+  funcOp.setAllArgAttrs(newArgAttrs);
 }
 
 //===----------------------------------------------------------------------===//
@@ -318,6 +286,7 @@ class PackDispatchOperandsPass
  public:
   void getDependentDialects(DialectRegistry &registry) const override {
     registry.insert<mlir::arith::ArithDialect>();
+    registry.insert<mlir::complex::ComplexDialect>();
     registry.insert<IREE::Stream::StreamDialect>();
   }
 

--- a/compiler/src/iree/compiler/Dialect/Stream/Transforms/Passes.cpp
+++ b/compiler/src/iree/compiler/Dialect/Stream/Transforms/Passes.cpp
@@ -287,8 +287,9 @@ void buildStreamOptimizationPassPipeline(
   // add/remove operands.
   passManager.addPass(IREE::Stream::createPackDispatchOperandsPass());
 
-  // Folding operands requires that CSE folds the inputs that we check for.
-  passManager.addPass(mlir::createCSEPass());
+  // Folding operands requires that canonicalization/CSE folds the inputs that
+  // we check for.
+  addCleanupPatterns(passManager);
   passManager.addPass(IREE::Stream::createFoldUniformOperandsPass());
 
   // Only want to specialize after we've added all the operands we need above.

--- a/compiler/src/iree/compiler/Dialect/Stream/Transforms/test/pack_dispatch_operands.mlir
+++ b/compiler/src/iree/compiler/Dialect/Stream/Transforms/test/pack_dispatch_operands.mlir
@@ -158,11 +158,11 @@ stream.executable private @ex4 attributes {stream.resources = #resourceIndex64} 
     // CHECK-SAME: (%[[DEV_LO32:.+]]: i32, %[[DEV_HI32:.+]]: i32, %{{.+}}: !stream.binding)
     func.func @device_index_64(%arg0: index {stream.alignment = 16 : index, stream.values = [0 : index, 1234 : index]}, %arg1: !stream.binding) {
       // 64-bit device size requires joining after it was split into lo/hi:
-      // CHECK: %[[DEV_LO64:.+]] = arith.extui %[[DEV_LO32]] : i32 to i64
-      // CHECK: %[[DEV_HI64:.+]] = arith.extui %[[DEV_HI32]] : i32 to i64
-      // CHECK: %[[DEV_HISHL:.+]] = arith.shli %[[DEV_HI64]], %c32
-      // CHECK: %[[DEV_I64:.+]] = arith.ori %[[DEV_LO64]], %[[DEV_HISHL]] : i64
-      // CHECK: %[[DEV_INDEX:.+]] = arith.index_castui %[[DEV_I64]] {
+      // CHECK-DAG: %[[DEV_LO64:.+]] = arith.extui %[[DEV_LO32]] : i32 to i64
+      // CHECK-DAG: %[[DEV_HI64:.+]] = arith.extui %[[DEV_HI32]] : i32 to i64
+      // CHECK-DAG: %[[DEV_HISHL:.+]] = arith.shli %[[DEV_HI64]], %c32
+      // CHECK-DAG: %[[DEV_I64:.+]] = arith.ori %[[DEV_LO64]], %[[DEV_HISHL]] : i64
+      // CHECK-DAG: %[[DEV_INDEX:.+]] = arith.index_castui %[[DEV_I64]] {
       // CHECK-SAME:   stream.alignment = 16 : index
       // CHECK-SAME:   stream.values = [0 : index, 1234 : index]
       // CHECK-SAME: } : i64 to index
@@ -187,6 +187,42 @@ func.func @host_index_64(%arg0: index) -> !stream.timepoint {
 
   %1 = stream.cmd.execute with(%0 as %arg1: !stream.resource<external>{%c128}) {
     stream.cmd.dispatch @ex4::@device_index_64[%c1, %c1, %c1](%arg0 : index) {
+      wo %arg1[%c0 for %c128] : !stream.resource<external>{%c128}
+    }
+  } => !stream.timepoint
+  return %1 : !stream.timepoint
+}
+
+// -----
+
+stream.executable private @ex5 {
+  // CHECK-LABEL: @device_complex_f32
+  stream.executable.export public @device_complex_f32
+  builtin.module {
+    // CHECK-LABEL: func.func @device_complex_f32
+    // CHECK-SAME: (%[[DEV_REAL_I32:.+]]: i32, %[[DEV_IMAG_I32:.+]]: i32, %arg2: !stream.binding)
+    func.func @device_complex_f32(%arg0: complex<f32>, %arg1: !stream.binding) {
+      // CHECK-DAG: %[[DEV_REAL_F32:.+]] = arith.bitcast %[[DEV_REAL_I32]] : i32 to f32
+      // CHECK-DAG: %[[DEV_IMAG_F32:.+]] = arith.bitcast %[[DEV_IMAG_I32]] : i32 to f32
+      // CHECK-DAG: %[[DEV_COMPLEX:.+]] = complex.create %[[DEV_REAL_F32]], %[[DEV_IMAG_F32]]
+      // CHECK-NEXT: util.optimization_barrier %[[DEV_COMPLEX]]
+      util.optimization_barrier %arg0 : complex<f32>
+      return
+    }
+  }
+}
+func.func @host_complex_f32(%arg0: complex<f32>) -> !stream.timepoint {
+  %c0 = arith.constant 0 : index
+  %c1 = arith.constant 1 : index
+  %c128 = arith.constant 128 : index
+  %0 = stream.resource.alloc uninitialized : !stream.resource<external>{%c128}
+  // CHECK-DAG: %[[HOST_REAL_F32:.+]] = complex.re %arg0
+  // CHECK-DAG: %[[HOST_IMAG_F32:.+]] = complex.im %arg0
+  // CHECK-DAG: %[[HOST_REAL_I32:.+]] = arith.bitcast %[[HOST_REAL_F32]] : f32 to i32
+  // CHECK-DAG: %[[HOST_IMAG_I32:.+]] = arith.bitcast %[[HOST_IMAG_F32]] : f32 to i32
+  %1 = stream.cmd.execute with(%0 as %arg1: !stream.resource<external>{%c128}) {
+    // CHECK: stream.cmd.dispatch {{.+}}(%[[HOST_REAL_I32]], %[[HOST_IMAG_I32]] : i32, i32)
+    stream.cmd.dispatch @ex5::@device_complex_f32[%c1, %c1, %c1](%arg0 : complex<f32>) {
       wo %arg1[%c0 for %c128] : !stream.resource<external>{%c128}
     }
   } => !stream.timepoint

